### PR TITLE
Add module OrderedMap

### PR DESCRIPTION
**Pull request by Rapiz1**
This module contains the implementation of the orderedMap type which is a container that stores key-value associations. 
OrderedSet supports searching for a certain key, insertion, and deletion in O(logN).

---

- [x] Drop the parameter `implType`
- [x] Migrate tests
- [x] Pass tests for OrderedSet

### CODE REVIEW

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:364]: Should this use the comparator to compare `result[0]` to `k`?

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:475]: `:type v:`

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:541]: Won't it be in order since `keys()` will yield the keys in order?

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:664]: Since you've written `|=` below, you can write this as (the equivalent but cleaner IMO)

```chapel
newMap |= a;
newMap |= b;
```

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:691]: This could use a note about how it invalidates references to every element in `a`

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl:241]: I tried many variants but none of them worked, including the way from Map you mentioned. I just deleted all types since their types are obvious. A wild guess is that it's a bug related to `var comparator: record = defaultComparator`. It's the only difference from Map I see.

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl:363]: Sounds reasonable

File: modules/packages/OrderedMap.chpl
Comment (cassella) [reply modules/packages/OrderedMap.chpl]: Possibly you could use `orderedMap(keyType, valType, ?)`?

I'm basing that on `doc/rst/technotes/partialInstantiations.rst`.

If that doesn't work, do you want to open an issue about it?

**Review by cassella — COMMENTED** [review]
A few more comments.

Is there a test that uses a comparator other than `defaultComparator`?  I'd like there to be one with a very different order so that it'll be clear if someone adds a direct `k1 < k2` instead of using the map's comparator.

(E.g. having the key be `record Element { var name: string, atomciNum: int, weight: real`; }, and provides a comparator that compares only by `atomicNum`.)

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl]: Removing all relevant lines since types are obvious.

File: test/library/packages/OrderedMap/types/mapOfArray.skipif
Comment (Rapiz1) [reply test/library/packages/OrderedMap/types/mapOfArray.skipif:3]: This is borrowed from Map. I don't know if it's necessary for orderedMap. But I supposed it's suitable for both map and orderedMap...

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl:155]: Hmm...I think `parSafe` is usually the parameter after types. However, I don't have strong opinions about this. @krishnadey30 @e-kayrakli What do you think?

File: test/library/packages/OrderedMap/getValue/testGetValue.good
Comment (Rapiz1) [reply test/library/packages/OrderedMap/getValue/testGetValue.good:1]: added in https://github.com/chapel-lang/chapel/pull/16271/commits/a13c9ad5e32a96071ba2e66c5a0a4acd69cd898d

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl]: https://github.com/chapel-lang/chapel/pull/16271/commits/b2949c093e02fb307f098a57e05508fae647d9b7

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:212]: I'd `no doc` this one.

File: modules/packages/OrderedMap.chpl
Comment (cassella) [reply modules/packages/OrderedMap.chpl:368]: Are you going to make that change?

**Review by cassella — COMMENTED** [review]
I'm nearly out of comments.  :)

File: modules/packages/OrderedMap.chpl
Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl:368]: Oh. Maybe I've clicked on this by accident. I will make the change

Comment (Rapiz1) [reply modules/packages/OrderedMap.chpl]: Updated

File: test/library/packages/OrderedMap/types/mapOfArray.skipif
Comment (cassella) [reply test/library/packages/OrderedMap/types/mapOfArray.skipif:3]: I just noticed though you're not adding a .future file for this test.

Map has the .future because valgrind reports a problem with that code.  It had that .skipif because the problem sometimes doesn't show up in multilocale (which start_test sees as a future that's now passing, which is something that requires attention.  So the skipif keeps it from running in cases where it might pass so that it only runs where it will fail.)

If the test works without valgrind errors for `OrderedMap`, then there'd be no reason to have a skipif.

I don't get any errors running the `OrderedMap` version of this test, with or without valgrind, with or without gasnet.  So I think you don't need the skipif.


(BTW, the Map test's skipif has been through a few changes since you copied it, `test/library/standard/Map/mapOfArray.skipif`.)

Comment (Rapiz1) [reply test/library/packages/OrderedMap/types/mapOfArray.skipif]: removed

File: modules/packages/OrderedMap.chpl
Comment (cassella) [inline modules/packages/OrderedMap.chpl:99]: Are these operators needed?  The `_keyComparator` below should be used anywhere that tuple needs to be compared (I think).  And all the tests pass with them removed.

Comment (cassella) [reply modules/packages/OrderedMap.chpl:99]: If they are needed internally to the module, I think they should be declared private, and/or `no doc`.  And probably there should be a test that fails if they're not present.

File: test/library/packages/OrderedMap/types/mapOfArray.skipif
Comment (cassella) [reply test/library/packages/OrderedMap/types/mapOfArray.skipif]: So marking the initial comment as resolved isn't what makes the thread resolved, and now I can't undo it.  :joy: 

Removing the skipif as you've done resolves this issue, but I still can't figure out how to get giithub to reflect that.

File: modules/packages/OrderedMap.chpl
Comment (cassella) [reply modules/packages/OrderedMap.chpl:156]: I dont' feel all that strongly about it, and I guess @krishnadey30 and @e-kayrakli don't either?

File: modules/packages/OrderedMap.chpl
Comment (cassella) [reply modules/packages/OrderedMap.chpl:99]: This could be taken care of after it's merged.

**Review by cassella — APPROVED** [review]
I think it would be better to merge this and potentially make the last two changes later than to let this languish any longer.

File: modules/packages/OrderedMap.chpl
Comment (e-kayrakli) [reply modules/packages/OrderedMap.chpl:156]: I find this version more appealing as `type` and `param` arguments come first in the argument list.

File: modules/packages/OrderedMap.chpl
Comment (e-kayrakli) [inline modules/packages/OrderedMap.chpl:166]: Today on master, this line is problematic: `_set` is declared as `orderedSet(_eltType, parSafe)` in the `orderedMap` definition. But here, you are assigning a `orderedSet(_eltType, false)` to it. This doesn't compile because of type mismatch. It may have been compiling before, but I think the compiler's behavior today is more correct.

The solution is to leave the `_set` s declaration a generic `orderedSet` and let the compiler figure out the right type for it.

I'm planning to do that change post-merge.

**Comment by e-kayrakli** [comment]
I am currently running testing on https://github.com/chapel-lang/chapel/pull/17450, which is supposed to fix a few issues with this PR. Except for `proc <`s that Paul noted to be unnecessary, all the failures are due to the changes in the compiler.

If that PR passes testing, I'll merge this, and that one right after.

**Pull request merged**
